File: frontend/src/services/teaching.ts
import type {
  BatchSpecAccepted,
  BatchSpecRequest,
  BatchSpecStatusPublic,
  CancelablePromise,
  ConfigPushAccepted,
  ConfigPushStatusPublic,
  HeatmapEntry,
} from "@/client"
import { OpenAPI } from "@/client"
import { request as __request } from "@/client/core/request"

const BASE = "/api/v1/teaching"

export const TeachingAPI = {
  getHeatmap(groupId: string): CancelablePromise<HeatmapEntry[]> {
    return __request(OpenAPI, {
      method: "GET",
      url: `${BASE}/heatmap`,
      query: { group_id: groupId },
    })
  },

  startConfigPush(opts: {
    file: File
    targetPath: string
    vmids: number[]
  }): CancelablePromise<ConfigPushAccepted> {
    return __request(OpenAPI, {
      method: "POST",
      url: `${BASE}/config-push`,
      formData: {
        file: opts.file,
        target_path: opts.targetPath,
        vmids: opts.vmids,
      },
    })
  },

  getConfigPushStatus(
    taskId: string,
  ): CancelablePromise<ConfigPushStatusPublic> {
    return __request(OpenAPI, {
      method: "GET",
      url: `${BASE}/config-push/{task_id}`,
      path: { task_id: taskId },
    })
  },

  startBatchSpec(body: BatchSpecRequest): CancelablePromise<BatchSpecAccepted> {
    return __request(OpenAPI, {
      method: "POST",
      url: `${BASE}/batch-spec`,
      body,
      mediaType: "application/json",
    })
  },

  getBatchSpecStatus(taskId: string): CancelablePromise<BatchSpecStatusPublic> {
    return __request(OpenAPI, {
      method: "GET",
      url: `${BASE}/batch-spec/{task_id}`,
      path: { task_id: taskId },
    })
  },
}

File: backend/app/api/routes/teaching.py
"""教學體驗 API（E2 配置分發 / E3 熱圖 / E6 批次規格），InstructorUser 起跳。"""

import asyncio
import logging
import uuid
from typing import Any

from fastapi import APIRouter, File, Form, UploadFile

from app.api.deps import InstructorUser, SessionDep
from app.infrastructure.proxmox import operations as proxmox_ops
from app.schemas import (
    BatchSpecAccepted,
    BatchSpecItemPublic,
    BatchSpecRequest,
    BatchSpecStatusPublic,
    ConfigPushAccepted,
    ConfigPushItemPublic,
    ConfigPushStatusPublic,
    HeatmapEntry,
)
from app.services.teaching import (
    batch_spec_service,
    config_push_service,
    progress_service,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/teaching", tags=["teaching"])


@router.post(
    "/config-push", response_model=ConfigPushAccepted, status_code=202
)
async def start_config_push(
    session: SessionDep,
    current_user: InstructorUser,
    file: UploadFile = File(...),
    target_path: str = Form(...),
    vmids: list[int] = Form(...),
) -> ConfigPushAccepted:
    content = await file.read()
    task_id = config_push_service.start_push(
        session,
        content=content,
        file_name=file.filename or "config",
        target_path=target_path,
        vmids=vmids,
        user=current_user,
    )
    return ConfigPushAccepted(task_id=task_id)


@router.get("/config-push/{task_id}", response_model=ConfigPushStatusPublic)
def get_config_push_status(
    task_id: str, current_user: InstructorUser
) -> ConfigPushStatusPublic:
    task = config_push_service.get_push_status(task_id, current_user)
    return ConfigPushStatusPublic(
        task_id=task.id,
        file_name=task.file_name,
        target_path=task.target_path,
        items=[
            ConfigPushItemPublic(vmid=i.vmid, status=i.status, error=i.error)
            for i in sorted(task.items.values(), key=lambda x: x.vmid)
        ],
    )


async def _safe_cluster_listing() -> list[dict[str, Any]]:
    try:
        return await asyncio.to_thread(proxmox_ops.list_all_resources)
    except Exception:
        logger.warning("Teaching: failed to list cluster resources", exc_info=True)
        return []


@router.get("/heatmap", response_model=list[HeatmapEntry])
async def get_heatmap(
    group_id: uuid.UUID,
    session: SessionDep,
    current_user: InstructorUser,
) -> list[HeatmapEntry]:
    cluster_resources = await _safe_cluster_listing()
    return progress_service.get_heatmap(
        session,
        group_id=group_id,
        user=current_user,
        cluster_resources=cluster_resources,
    )


@router.post("/batch-spec", response_model=BatchSpecAccepted, status_code=202)
def start_batch_spec(
    body: BatchSpecRequest,
    session: SessionDep,
    current_user: InstructorUser,
) -> BatchSpecAccepted:
    task_id = batch_spec_service.start_batch_spec(
        session,
        vmids=body.vmids,
        group_id=body.group_id,
        cores=body.cores,
        memory_mb=body.memory_mb,
        user=current_user,
    )
    return BatchSpecAccepted(task_id=task_id)


@router.get("/batch-spec/{task_id}", response_model=BatchSpecStatusPublic)
def get_batch_spec_status(
    task_id: str, current_user: InstructorUser
) -> BatchSpecStatusPublic:
    task = batch_spec_service.get_batch_status(task_id, current_user)
    return BatchSpecStatusPublic(
        task_id=task.id,
        items=[
            BatchSpecItemPublic(vmid=i.vmid, status=i.status, error=i.error)
            for i in sorted(task.items.values(), key=lambda x: x.vmid)
        ],
    )
